server/server.go

```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
 	}
 	return listeners
 }
+
+// Returns true if in lame duck mode.
+func (s *Server) isLameDuckMode() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.ldm
+}
+
+// This function will close client and route listeners
+// then close clients at some interval to avoid a reconnecting storm.
+// Note that the routes are closed before so that when a client is
+// disconnected and reconnects to another server, that server does
+// not send to this server a route SUB.
+func (s *Server) lameDuckMode() {
+	s.mu.Lock()
+	// If we are in the process (or done) or if listener is not yet
+	// ready, just return.
+	if s.ldm || s.listener == nil {
+		s.mu.Unlock()
+		return
+	}
+	expected := 1
+	s.Noticef("Entering lame duck mode, stop accepting new clients")
+	s.ldm = true
+	s.ldmCh = make(chan bool, 1)
+	s.ldmDoneCh = make(chan bool, 1)
+	s.listener.Close()
+	s.listener = nil
+	if s.routeListener != nil {
+		s.routeListener.Close()
+		s.routeListener = nil
+		expected++
+	}
+	s.mu.Unlock()
+
+	// Wait for accept loop to be done to make sure that no new
+	// client can connect
+	for i := 0; i < expected; i++ {
+		<-s.ldmCh
+	}
+
+	durInMs := int64(s.getOpts().LameDuckDuration / time.Millisecond)
+	if durInMs < 1 {
+		durInMs = 1
+	}
+
+	// Close routes
+	s.mu.Lock()
+	routes := make([]*client, 0, len(s.routes))
+	for _, r := range s.routes {
+		routes = append(routes, r)
+	}
+	s.mu.Unlock()
+	for _, r := range routes {
+		r.setRouteNoReconnectOnClose()
+		r.closeConnection(LameDuckMode)
```

Comment:
I think this is not what we want since it means clients who are still connected and communicating won't work.

Reply:
If you are ok with the fact that when client reconnect to another server this server starts to receive route SUBs, then I can remove that.

Reply:
I think for now that is ok. Say for example we have simple pub and sub connected to LD server, we migrate the subscriber. We want the PUB to succeed as best as possible. What might be good though if possible is to have the LD server let all the other servers to not advertise it as an option for reconnect.

Reply:
Removed closing of routes.

Reply:
Also, I think we don't need complexity of communicating between server that server is in LD mode. Sine the LD server will have shutdown its listener port, even if client try to reconnect to it, they will fail and move to other server.